### PR TITLE
[Silabs][SiWx917] Factory data provider for SiWx917

### PR DESCRIPTION
**Pull request by chirag-silabs**
Factory data provider for SiWx917
This is a temp fix, the user can write the details in the DeviceConfig.h (Discriminator, passcode and all) and then build the project with `siwx917_commissionable_data=true` and then the device properties will be flashed in the nvm section of the SiWx917 board.

### CODE REVIEW

File: examples/platform/silabs/SiWx917/SiWx917DeviceDataProvider.cpp [hunk 1]
```diff
@@ -24,12 +26,183 @@
...
+            bits[int(index / 8)] |= (1 << (index % 8));
+        }
+        index++;
+        input >>= 1;
+    }
+    return offset;
+}
+
+// Generation of the payload using the details provided in the DeviceConfig.h
+void SIWx917DeviceDataProvider::generateQrCodeBitSet(uint8_t * payload)
```
Comment (bzbarsky-apple) [inline examples/platform/silabs/SiWx917/SiWx917DeviceDataProvider.cpp:59]: Is there a reason this is not using the existing payload generator we have in the tree?

Comment (chirag-silabs) [reply examples/platform/silabs/SiWx917/SiWx917DeviceDataProvider.cpp:59]: Actually this writes in byte by byte format, and I was getting incorrect value by payload generation present in the tree. I was getting the payload value directly (MT:xxxxxxxxxxxxxxxxxxx) and this provides in byte by byte format and flashes that in the NVM section.

**Review by jmartinez-silabs — APPROVED** [review]

File: examples/platform/silabs/SiWx917/SiWx917DeviceDataProvider.cpp (hunk 1 above)
Comment (bzbarsky-apple) [reply examples/platform/silabs/SiWx917/SiWx917DeviceDataProvider.cpp:59]: @chirag-silabs I'm not following the distinction here.  Is the issue that you want to exclude the "MT:" bit?  Or that you want to not do the base-38 encoding?  Or something else?

My instinct is that duplicating the bit-pushing here is very failure-prone and I would really like to understand what prevents us from using the existing code we have that does that and has tests and whatnot.

Comment (chirag-silabs) [reply examples/platform/silabs/SiWx917/SiWx917DeviceDataProvider.cpp]: @bzbarsky-apple address the comments on the latest commit

Comment (bzbarsky-apple) [inline examples/platform/silabs/SiWx917/SiWx917DeviceDataProvider.cpp:59]: Again, why are we not using our existing payload generator functionality?  The previous thread claims "address the comments on the latest commit" but the code has not particularly changed, and the questions I asked were not answered.

**Review by bzbarsky-apple — CHANGES_REQUESTED** [review]

File: examples/platform/silabs/SiWx917/SiWx917DeviceDataProvider.cpp (hunk 1 above)
Comment (chirag-silabs) [reply examples/platform/silabs/SiWx917/SiWx917DeviceDataProvider.cpp]: Sorry my bad, When I rebased the branch I lost the commit which was done to address the comments. Pushed the latest commit

File: examples/platform/silabs/SiWx917/SiWx917DeviceDataProvider.cpp [hunk 2]
```diff
@@ -16,20 +16,158 @@
...
+void SIWx917DeviceDataProvider::setupPayload(uint8_t * outBuf)
+{
+    SetupPayload payload;
+    std::string result;
+    ChipError err   = CHIP_NO_ERROR;
+    payload.version = 0;
+    payload.discriminator.SetLongValue(discriminatorValue);
+    payload.setUpPINCode = passcode;
+    payload.rendezvousInformation.SetValue((chip::RendezvousInformationFlags) rendezvousFlag);
+    payload.commissioningFlow = (chip::CommissioningFlow) commissionableFlow;
```
Comment (bzbarsky-apple) [inline examples/platform/silabs/SiWx917/SiWx917DeviceDataProvider.cpp:46]: Please use static_cast, not C-style casts.

File: examples/platform/silabs/SiWx917/SiWx917DeviceDataProvider.cpp [hunk 3]
```diff
@@ -16,20 +16,158 @@
...
+    }
+    SILABS_LOG("Payload value in string format : %s", result.c_str());
+    std::vector<uint8_t> result1;
+    // skipping the MT: from the payload during decoding
+    err = base38Decode(result.substr(3), result1);
+    if (err == CHIP_NO_ERROR)
+    {
+        for (uint8_t i = 0; i < result1.size(); i++)
+        {
+            outBuf[i] = result1.at(i);
```
Comment (bzbarsky-apple) [inline examples/platform/silabs/SiWx917/SiWx917DeviceDataProvider.cpp:64]: The lack of a length check here is pretty fragile, but ok.

**Review by bzbarsky-apple — APPROVED** [review]

**Review by rcasallas-silabs — APPROVED** [review]

File: examples/platform/silabs/SiWx917/SiWx917DeviceDataProvider.cpp (hunk 3 above)
Comment (andy31415) [reply examples/platform/silabs/SiWx917/SiWx917DeviceDataProvider.cpp:64]: Resolving to merge - platform owner approved this potential out of bounds access :/

Comment (andy31415) [reply examples/platform/silabs/SiWx917/SiWx917DeviceDataProvider.cpp:64]: Merged, now re-opening for visibility. The method should have taken a span as input to do bounds checking.

**Pull request merged**